subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/dsl/DefaultRepositoryHandler.java

```diff
@@ -146,4 +160,94 @@ public IvyArtifactRepository ivy(Action<? super IvyArtifactRepository> action) {
+                    desc.excludeModule(group, moduleName);
+                }
+
+                @Override
+                public void includeModuleByRegex(String groupRegex, String moduleNameRegex) {
+                    desc.excludeModuleByRegex(groupRegex, moduleNameRegex);
+                }
+
+                @Override
+                public void includeVersion(String group, String moduleName, String version) {
+                    desc.excludeVersion(group, moduleName, version);
+                }
+
+                @Override
+                public void includeVersionByRegex(String groupRegex, String moduleNameRegex, String versionRegex) {
+                    desc.excludeVersionByRegex(groupRegex, moduleNameRegex, versionRegex);
+                }
+            });
+        };
+    }
+
+    public static class ExclusiveContentRepositorySpec implements ExclusiveContentRepository {
+        private final RepositoryHandler repositories;
+        private final List<Factory<? extends ArtifactRepository>> forRepositories = Lists.newArrayListWithExpectedSize(2);
+        private Action<? super InclusiveRepositoryContentDescriptor> filter;
+
+        public ExclusiveContentRepositorySpec(RepositoryHandler repositories) {
+            this.repositories = repositories;
+        }
+
+        @Override
+        public ExclusiveContentRepository forRepository(Factory<? extends ArtifactRepository> repositoryProducer) {
+            forRepositories.add(repositoryProducer);
+            return this;
+        }
+
+        @Override
+        public ExclusiveContentRepository forRepositories(ArtifactRepository... repositories) {
+            Stream.of(repositories).forEach(r -> forRepositories.add(() -> r));
+            return this;
+        }
+
+        @Override
+        public ExclusiveContentRepository filter(Action<? super InclusiveRepositoryContentDescriptor> config) {
+            filter = filter == null ? config : Actions.composite(filter, config);
+            return this;
+        }
+
+        void apply(Action<? super ExclusiveContentRepository> action) {
+            action.execute(this);
+            if (forRepositories.isEmpty()) {
+                throw new InvalidUserCodeException("You must declare the repository using forRepository { ... }");
+            }
+            if (filter == null) {
+                throw new InvalidUserCodeException("You must specify the filter for the repository using filter { ... }");
+            }
+            Set<? extends ArtifactRepository> targetRepositories = forRepositories.stream().map(Factory::create).collect(Collectors.toSet());
+            Action<? super RepositoryContentDescriptor> forExclusivity = transformForExclusivity(filter);
+            this.repositories.all(repo -> {
+                if (targetRepositories.contains(repo)) {
+                    repo.content(filter);
+                } else {
+                    repo.content(forExclusivity);
+                }
+            });
```

Comment:
Does this support the removal of the repository?

In other words, what if someone decides to remove a repository or remove `exclusiveContent` later?

Reply:
No it doesn't. It's a good point, although removing repositories is possible, it's not something recommended.